fixed unzip bug when files and folders have the same name for overriding

hyi opened the pull request:
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [ ] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]

hyi commented:
@devincowan @pkdash This PR should fix both unzip bugs #4752 and #4753. Please code review this PR when you get a chance so that next release can proceed. Thanks

hydrocheck commented:
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4754](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4754)

devincowan commented:
@hyi thanks for the quick turn around on this!
I see that your modifications are to 
[utils.unzip_file()](https://github.com/hydroshare/hydroshare/blob/4752-unzip-bug/hs_core/views/utils.py#L1036)

However, I believe the function that gets called during unzipping in place is
[resource_folder_hierarchy.data_store_folder_unzip]https://github.com/hydroshare/hydroshare/blob/4752-unzip-bug/hs_core/views/resource_folder_hierarchy.py#L445

I just want to make sure we are working on the correct function before we dive into the weeds...thanks!

hyi commented:
@devincowan Yes, utils.unzip_file() is the method being called by [resource_folder_hierarchy.data_store_folder_unzip()](https://github.com/hydroshare/hydroshare/blob/4752-unzip-bug/hs_core/views/resource_folder_hierarchy.py#L445) indirectly. Anyway, this is the correct function that actually does the main work. It'd be great if you could pull my bug fix branch and test it locally to verify whether everything works as expected. Thanks

devincowan commented:
ah I see that now @hyi thanks! I'll take a look

devincowan commented:
@hyi thanks for keeping me on track ;)
Looks like this PR fixes #4752 
However I still see #4753 on my local build after creating a bagit zip, uploading it, and unzipping in place.

hyi commented:
@devincowan I could not test issue #4753 since I am not sure about the Step 2: Create a zip in the content browser. I tagged @ndebuhr but has not got his response yet. Since you can reproduce the issue #4753, can you let me know what the step #2 meant in the steps to reproduce the bug or let me know if you have a simplified steps to reproduce the issue?

devincowan commented:
@hyi I will DM you some info
@pkdash is there any way that the exception in #4753 (which is looking for readme.txt but accidentally finding the nested readme within the unzipped bagit) could be related to your
PR #4711 
Specifically some of these resource.files modifications?
https://github.com/hydroshare/hydroshare/pull/4712/files#diff-a22b5d7504c86ddd514ee34a21d926c083f44344e94bcc580900e0ef368d1550R62

pkdash commented on hs_core/views/utils.py:
```diff
@@ -1072,10 +1072,16 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
+        for sub_dir_name in unzip_subdir_list:
+            dest_sub_dir_path = os.path.join(os.path.dirname(zip_with_full_path), sub_dir_name)
+            if istorage.exists(dest_sub_dir_path):
+                override_tgt_paths.append(dest_sub_dir_path)
```
I think we may have to have separate list for folders that will be overridden. That way we can first loop though that folder list first and delete folders before deleting the files that would be overridden.

hyi replied:
I tried to make changes as you suggested since your comment made sense to me. However, after doing so, the bug was not fixed any more. After debugging, I realized the issue was related to a folder overriding a file with the same name, so we cannot really use the list for folders for overriding to determine whether we need to remove folder or file from the destination folder due to the fact that a folder with the same name as a file in the destination folder cannot be moved to the destination folder until the file with the same name is removed. The reverse is true when moving a file to a folder with the same name, so the user has to determine whether they want to really override either files or folders with the same name as the files or folders to be unzipped. It is really confusing, but I have to go back to the old logic without splitting into folders and files to get this working.

pkdash commented on hs_core/views/utils.py:
```diff
@@ -1100,7 +1106,19 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
+                                # delete all files in the folder
+                                files.delete()
+                            # delete the folder in iRODS
+                            istorage.delete(override_tgt_path)
```
I think this line which deletes the folder will also delete all files in that folder. In the `else` block only this line should be sufficient. Also we need to use `remove_folder` function from `views/utils.py` to delete the folder which will take care of any aggregations associated with that folder.

hyi replied:
Good point. Thanks for the suggestion.

pkdash reviewed (commented):
Left some comments for your consideration. Do we have unit tests for unzip functionality implemented here?

hyi commented:
To clarify, I did not get unzip failure, the unzip actually completed successfully, however, when I clicked the link to refresh resource landing page, Multiple Objects returned exception resulted so the resource landing page could not be displayed successfully. I ended up deleting the resource and created a new resource using a zip file that does not included nested readme.md. So it appears to be an issue with nested readme.md, i.e., multiple readme.md included in the same resource in the nested folders, which appears to result in issues for resource landing page display, which appears to be a separate issue from the unzip action.

hyi commented on hs_core/views/utils.py:
```diff
@@ -1100,7 +1106,19 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
                     else:
-                        istorage.delete(override_tgt_path)
+                        # check if destination override path is a file
+                        f = ResourceFile.get(resource=resource, file=override_tgt_path)
```
Good point. I'll add catch exception here.

hyi commented:
@pkdash I have addressed all your code review comments and fixed a couple of additional issues related to file overriding folder with the same name and also made the code structure a bit better and easier to read. There are tests that covered unzipping in place, but no specific tests to test this file overriding folder and folder overriding test issues that this PR should have fixed, so I created a new issue #4758 to work on adding tests later in the interest of getting the bug fixes into release 1.58 in time. This PR should ready for your review again. @devincowan It'd be good for you to test this out as well since I verified the unzip fix in this PR along with Pabitra's readme.md fix PR will indeed fix the resource broken issue #4753. I was able to reproduce the resource broken issue #4753 without getting the readme.md fix in, and after getting the readme.md fix in, the resource landing page just works. So the broken resource exception was triggered by the readme.md display issue and with Pabitra's fix along with my unzip in place fix in this PR, both bugs blocking release 1.58 should be fixed. I tested all kinds of unzip workflows I can think of and everything works, but it'd be great if you and Pabitra can do a test as well to make sure no new regressions would be introduced. Thanks

devincowan reviewed (approved):
Looks good to me.
Thanks for the fast turn-around
There is one additional issue that Pabitra discovered that we will need to address before the release...
